src/fetch_agent/fetch_agent/sort_stations.py

```diff
@@ -0,0 +1,226 @@
+    ctx: Context,
+    serialized: list[Tuple[str, PropertyData]],
+    save_name: str = json_key_stations_properties_map,
+):
+    ctx.logger.info(
+        f"[Sort Stations, serialize_and_save_stations_properties_map]: Starting serializing and saving to {save_name}"
+    )
+    to_save: list[Tuple[str, str]] = []
+
+    for item in serialized:
+        to_save.append((item[0], item[1].to_json()))
+
+    ctx.storage.set(save_name, to_save)
+
+
+def _read_car_properties(ctx: Context) -> PropertyCarData:
+    ctx.logger.info("[Sort Stations, read_car_properties]: Starting reading")
+
+    return PropertyCarData(
+        green_energy=ctx.storage.get("green_energy_weight"),
+        cost_per_kwh=ctx.storage.get("cost_per_kwh_weight"),
+        charging_wattage=ctx.storage.get("charging_wattage_weight"),
+        max_km=ctx.storage.get("filter_max_km"),
+        time_frames=ctx.storage.get("open_time_frames"),
+    )
+
+
+def initialize_stations_properties_map(ctx: Context):
+    ctx.storage.set(json_key_stations_properties_map, [])
+
+
+def set_PropertyData_of_sender(ctx: Context, sender: str, properties: str):
+    if ctx.storage.get("finished_waiting"):
+        ctx.logger.info(
+            f"[Sort Stations, set_PropertyData_of_sender]: Ignoring properties of {sender}"
+        )
+        return
+
+    ctx.logger.info(
+        f"[Sort Stations, set_PropertyData_of_sender]: Setting properties of {sender} to {properties}"
+    )
+
+    stations_properties: list[Tuple[str, str]] = ctx.storage.get(
+        json_key_stations_properties_map
+    )
+
+    stations_properties.append((sender, properties))
+
+    ctx.storage.set(json_key_stations_properties_map, stations_properties)
+
+
+def initialize_car_properties(ctx: Context):
+    ctx.storage.set("car_geo_point", (20.32, 85.52))
+
+    if not ctx.storage.get("green_energy_weight"):
+        ctx.storage.set("green_energy_weight", 1)
+
+    if not ctx.storage.get("cost_per_kwh_weight"):
+        ctx.storage.set("cost_per_kwh_weight", 5)
+
+    if not ctx.storage.get("charging_wattage_weight"):
+        ctx.storage.set("charging_wattage_weight", 1)
+
+    if not ctx.storage.get("filter_max_km"):
+        ctx.storage.set("filter_max_km", 20)
+
+    if not ctx.storage.get("open_time_frames"):
+        ctx.storage.set(
+            "open_time_frames",
+            [
+                (
+                    int((datetime.now() + timedelta(hours=1)).timestamp()),
+                    int((datetime.now() + timedelta(hours=4)).timestamp()),
+                ),
+            ],
+        )
+
+
+def sort_stations(ctx: Context) -> (str, PropertyData, Tuple[int, int]):
+    stations_properties = _read_stations_properties_map(ctx)
+
+    if not stations_properties:  # case no station responded
+        ctx.logger.warning(
+            f"[Sort Stations, sort_stations]: No station responded. Here is nothing to filter: {stations_properties}. So there is no optimal station"
+        )
```

Comment:
error message is not very good
it does not necessarly mean that no station responded